ao-tools/ao-dumpflash/ao-dumpflash.1
.\"
.\" Copyright © 2013 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-DUMPFLASH 1 "ao-dumpflash" ""
.SH NAME
ao-dumpflash \- Fetch flash memory contents from AltOS device
.SH SYNOPSIS
.B "ao-dumpflash"
[\--tty \fItty-device\fP]
[\--device \fIaltos-device\fP]
[\--output \fIoutput-file\fP]
[\--remote\fP]
[\--frequency \fIfrequency\fP]
[\--call \fIcallsign\fP]
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device ao-dumpflash uses to communicate with
the target device.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMetrum:2
.br
TeleMetrum
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.TP
\-o output-file | --output output-file
Write flash contents to the specified file rather than stdout.
.TP
\-R | --remote
This uses the command radio link to download the flash from TeleMetrum
through a TeleDongle.
.TP
\-F frequency | --frequency frequency
Specifies the radio frequency to use for remote communications in
kHz. Default is 434550.
.TP
\-C callsign | --call callsign
Specifies the callsign to use for remote communications. Default is N0CALL.
.SH DESCRIPTION
.I ao-dumpflash
downloads the entire flash memory contents from a connected AltOS device and writes
it to either stdout or the specified output file.
.SH USAGE
.I ao-dumpflash
connects to the specified target device and dumps the flash.
.SH AUTHOR
Keith Packard


ChangeLog
commit 0cd443d5e55b2c3b97ecf53389ff76bc4bc6018f
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 16 12:38:33 2012 -0700

    Add release notes for version 1.1.1
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit cb48dd0b03b445437f751028e8383610b65b0a68
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 16 02:39:25 2012 -0700

    altos: Fix telemini sdcdb init file
    
    Set the driver source path for the debugger
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3520bbf1ed6461d1ce7af001c529563a3cffa3c9
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Sep 16 00:58:20 2012 -0700

    altosui: Gather Tm sensor data in Monitor Idle mode
    
    Tm is pretty much the same as TM for the analog sensors, it's just
    missing the accelerometr. Use the same code for constructing an
    AltosRecord for it
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 31f5a02654cbf172beed25f4c518dfb7be8c714e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 16:56:04 2012 -0700

    altos: Need to use 16-bit counts for ao_xmem functions
    
    Trying to use 8-bit counts is a nice optimization which fails when the
    count is larger than 255, as is the case with clearing the flash block
    in the AT45 driver. This bug resulted in the inability to erase
    flights on TeleMetrum v1.0 boards.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b898cf0a2abf2b0478d5afc5aca030c6b4c8bd0b
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 12:59:31 2012 -0700

    altosui: Show over-ground-distance in Descent tab
    
    Helps to know where the rocket might land.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f36f73b2d02b72201683cf5795851034bbd6f28e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 12:58:54 2012 -0700

    altosui: Remove debugging from KML export
    
    Left in from testing the Eeprom export bug
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 97ab77d548964115e4b41ad5952194fcd1455c96
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 11:13:02 2012 -0700

    altosui: Fix Landed tab units
    
    And clean up the whole flight value reporting code base. It would be
    nice to create a separate class to make this easier; at present
    there's a bunch of customization embedded in how values are presented
    in each tab.
    
    Reported by: Bdale Garbee <bdale@gag.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 43e646657502f6162fa02f37fd2bd5aa3e29a1a8
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 11:08:57 2012 -0700

    altoslib: Update GPS seen_values in AltosEepromIterable
    
    Otherwise, KML export won't ever get GPS data to write.
    
    Reported by: Bob Brown <prefect@kloudbusters.org>
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7b6dd9105ba36aa11d6d0ee6e0823965b9beffb9
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Sep 14 11:08:19 2012 -0700

    altoslib: Remove duplicate seen_ values in AltosEepromIterable
    
    These are defined in AltosRecord; duplicating them is not a good plan
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit eefcfa94f360f8c5a7233370d4178525bccbb22f
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Sep 13 16:13:42 2012 -0600

    add note about checking doc copyright date and revision history to Releasing

commit 2441090f26c3df66a6ce48d64e64384ce9e5ef82
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Sep 13 15:12:33 2012 -0700

    doc: Add revision history for 1.1. Update copyright date
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 320d90c376dccfe1599505e3b485df8d46e34bb3
Merge: 9a7d643 2e6c6a6
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Sep 13 15:36:37 2012 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 2e6c6a6c15a7cff9f20b70a4cf58697da1302f01
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Sep 13 00:35:27 2012 -0700

    doc: Document imperial units and groundstation configuration
    
    What it says on the tin.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9a7d6431777ce3377b788ddac6cb9fadd53c039c
Merge: 2439f53 9728b20
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Sep 13 00:58:30 2012 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 9728b20a4ad9576ac1022dc9a5771f5a3eb92b3d
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 12 23:56:18 2012 -0700

    altosui: Exit with an error status when file processing fails
    
    If the user provides any files on the command line, and if processing
    them fails in some way, exit immediately with an error indication.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0564f9b13d4e3050b0a45fc1c4f710ab2b1f9d81
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 12 22:40:57 2012 -0700

    altosui: Leave new_gps indication until gps value is used
    
    During replay, AltosState may not see a new GPS value as soon as it
    lands in the state field as additional records with the same timestamp
    may come in after the GPS record.
    
    Instead of resetting the new_gps indication when the new record is
    created, wait until the new_gps indication is seen by the AltosState
    update code and have that clear the new_gps indication.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7e237920abb582d7ceb3e1925b11a848575ac68a
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Sep 12 21:32:25 2012 -0700

    altosui: Stop downloading eeprom data on a block full of invalid data
    
    When no valid records are found within an eeprom block, we assume that
    no more data will be found within the entire storage area.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2439f53ea21c84f7c510dc191dde025ada24281f
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Sep 12 21:50:25 2012 -0600


ao-tools/ao-dumpflash/ao-dumpflash.c
/*
 * Copyright © 2009 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include "cc-usb.h"
#include "cc.h"

#define NUM_BLOCK	512

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "remote", .has_arg = 0, .val = 'R' },
	{ .name = "frequency", .has_arg = 1, .val = 'F' },
	{ .name = "call", .has_arg = 1, .val = 'C' },
	{ .name = "output", .has_arg = 1, .val = 'o' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--tty <tty-name>] [--device <device-name>] [--remote] [--frequency <radio-frequency>] [--call <radio-callsign>]\n", program);
	exit(1);
}

int
main (int argc, char **argv)
{
	struct cc_usb	*cc;
	char		*tty = NULL;
	char		*device = NULL;
	int		c;
	char		line[8192];
	FILE		*out;
	char		*filename;
	int		serial_number = 0;
	int		freq = 434550;
	char		*call = "N0CALL";
	int		flight = 0;
	char		cmd;
	int		block;
	int		addr;
	int		received_addr;
	int		data[8];
	int		done;
	int		i;
	int		column;
	int		remote = 0;
	int		any_valid;
	int		invalid;
	int		storage_size = 0;
	char		*out_name;

	while ((c = getopt_long(argc, argv, "T:D:F:C:o:R", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'R':
			remote = 1;
			break;
		case 'F':
			freq = atoi(optarg);
			break;
		case 'C':
			call = optarg;
			break;
		case 'o':
			out_name = optarg;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	if (!tty) {
		if (remote)
			tty = cc_usbdevs_find_by_arg(device, "TeleDongle");
		else
			tty = cc_usbdevs_find_by_arg(device, "TeleMetrum");
	}
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyACM0";

	cc = cc_usb_open(tty);
	if (!cc)
		exit(1);
	if (remote)
		cc_usb_open_remote(cc, freq, call);

	if (out_name) {
		out = fopen(out_name, "w");
		if (!out) {
			perror(out_name);
			cc_usb_close(cc);
			exit(1);
		}
	} else
		out = stdout;

	/* send a 'version' command followed by a 'flash' command */
	cc_usb_printf(cc, "f\nv\n");
	for (;;) {
		cc_usb_getline(cc, line, sizeof (line));
		if (sscanf(line, "serial-number %u", &serial_number) == 1)
			continue;
		if (sscanf(line, "Storage size: %u", &storage_size) == 1)
			continue;
		if (!strncmp(line, "software-version", 16))
			break;
	}
	if (!serial_number) {
		fprintf(stderr, "no serial number found\n");
		cc_usb_close(cc);
		exit(1);
	}
	if (!storage_size) {
		fprintf(stderr, "no storage size found\n");
		cc_usb_close(cc);
		exit(1);
	}
	printf ("Serial number: %d\n", serial_number);
	printf ("Storage size:  %d\n", storage_size);
	fprintf (stderr, "%7d of %7d", 0, storage_size/256);
	for (block = 0; block < storage_size / 256; block++) {
		cc_usb_printf(cc, "e %x\n", block);
		fprintf (stderr, "\r%7d of %7d", block + 1, storage_size/256); fflush(stderr);
		for (addr = 0; addr < 0x100;) {
			cc_usb_getline(cc, line, sizeof (line));
			if (sscanf(line, "00%x %x %x %x %x %x %x %x %x",
					  &received_addr,
					  &data[0], &data[1], &data[2], &data[3],
					  &data[4], &data[5], &data[6], &data[7]) == 9)
			{
				if (received_addr != addr)
					fprintf(stderr, "data out of sync at 0x%x\n",
						block * 256 + received_addr);

				fprintf (out, "%08x", block * 256 + addr);
				for (i = 0; i < 8; i++)
					fprintf (out, " %02x", data[i]);
				fprintf (out, "\n");

				addr += 8;
			}
		}
	}
	fprintf(stderr, "\n");
	cc_usb_close(cc);
	exit (0);
}
